### ISSUE TITLE
[regression - rust2018]: unused_mut lint false positives on nightly

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#55344]
I believe this is a regression in rust2018 nightly (it wasn't triggering in earlier rust2018 builds).

---

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=efcfd792e59b4164532d6b11976b11d5)

```rust
#![deny(warnings)]  

#[allow(unreachable_code)]
pub fn sum_nan() {
    // return;

    let mut v = 0;
    assert_eq!(v, 0);
    v = 1;
    assert_eq!(v, 1);
}

fn main() {}
```

works correctly, but uncommenting the `return;` ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f4d21b150d16363b1d396f5c0d62a52e)) errors with: 

```
error: variable does not need to be mutable
 --> src/main.rs:7:9
  |
7 |     let mut v = 0;
  |         ----^
  |         |
  |         help: remove this `mut`
```

This only happens with Rust2018 on nightly (Rust2015 works fine).

**Comment by matthewjasper** [comment]
Well, the code compiles (in Rust2018) without `mut`, so the lints not exactly wrong, even if the behaviour is surprising.

**Comment by gnzlbg** [comment]
@matthewjasper accepting those programs without the mut in Rust 2018 or with the `nll` feature in Rust2015 appears to be a soundness bug: https://github.com/rust-lang/rust/issues/54943 (the programs are illegal and are incorrectly accepted by the compiler)

**Comment by pnkfelix** [comment]
#54943 is about well-formedness, which I don't think is part of the code that tags whether a `mut` was necessary or not (which is what is triggering the `unused_mut` warning that is then triggering the `deny(warnings)` you have in there.

This is an interesting scenario in terms of what one can do to mitigate this...

**Comment by gnzlbg** [comment]
@pnkfelix i've been thinking a bit about the well-formedness of unreachable code because of this. AFAICT with NLL currently unreachable code needs to be syntactically correct, to `typeck` and to `unsafeck`, but it does not need to `dropck` (double drops are currently allowed) or `borrowck` (this example).

In Rust2015 unreachable code also needs to `dropck` and `borrowck`.

**Comment by gnzlbg** [comment]
The first question that should be answered is whether this change in the well-formedness requirements of unreachable code are by design or an accident (cc @nikomatsakis ). If this is by design, then I'd suppose they have gone through an RFC and we can close this, or leave this open with the intent of maybe adding a better warning about this. 

If they are by accident, then the first thing to do is to decide what should be the correct behavior here (e.g. stick to what Rust2015 does or change that to something else?).

**Comment by nikomatsakis** [comment]
This is considered an acceptable regression, per our usual policy on lints (see e.g. the motivation for [RFC 1993](https://rust-lang.github.io/rfcs/1193-cap-lints.html). However, we may want to avoid linting about unused-mut on unreachable variables like this, just to be less annoying.

**Comment by pnkfelix** [comment]
(in particular the fact that the *binding* is itself unreachable makes this lint seem especially weird, at least to me...)

**Comment by pnkfelix** [comment]
Also, I am not sure whether I was sufficiently clear in my [earlier comment](https://github.com/rust-lang/rust/issues/55344#issuecomment-434122225): #54943 is using the phrase "well-formed" as a particular technical term. "Well-formed" is a predicate on types that is meant to evaluate whether an individual type "makes sense" before you start considering whether other expressions are instances of that type;  it is documented (at least in part) in part of [RFC 1214](https://rust-lang.github.io/rfcs/1214-projections-lifetimes-and-wf.html#well-formed-types). 

In other words, I do not think #54943 is related to this issue.

**Comment by gnzlbg** [comment]
Another variant of this bug is:

```
#[allow(unreachable_code)]
pub fn sum_nan() {
    return;

    let v = 0;
    assert_eq!(v, 0);
    std::mem::drop(v);
    std::mem::drop(v);
    assert_eq!(v, 1);
}
```

where no errors are emitted even though the unreachable code has a double drop, or this code:

```rust
#[allow(unreachable_code, unused_mut)]
pub fn sum_nan() {
    return;

    let mut v = 0;
    assert_eq!(v, 0);
    let x = &mut v;
    assert_eq!(*x, 0);
    *x = 1;
    let y = &mut v;
    assert_eq!(*y, 1);
    assert_eq!(*x, 1);
}
```

where two mutable references can be created and compile fine in unreachable code.

**Comment by pnkfelix** [comment]
@gnzlbg [wrote](https://github.com/rust-lang/rust/issues/55344#issuecomment-434202940):

> The first question that should be answered is whether these changes in the well-formedness requirements of unreachable code are by design or an accident (cc @nikomatsakis ). If this is by design, then I'd suppose they have gone through an RFC and we can close this, or leave this open with the intent of maybe adding a better warning about this. 

I think at least *some* of the changes you are describing here are definitely by design.

The [NLL RFC](https://github.com/rust-lang/rfcs/blob/master/text/2094-nll.md) stated that lifetimes were going to be based on the control-flow graph. This included details like "the lifetime `'a` must only outlive those portions of `'b` that are reachable from P."

Giiven that this was always going to be a control-flow sensitive analysis, it seems entirely understandable that the new borrow-check analysis is going to only consider code that is actually reachable in the control flow.

So, my initial reaction is that the examples you describe are not all that surprising to me.

Having said that ... that same RFC *did* document a number of changes that were being included in order to [deal with infinite loops](https://github.com/rust-lang/rfcs/blob/master/text/2094-nll.md#layer-2-avoiding-infinite-loops).

 * (A stray thought: Would it be bonkers to use similar false unwind edges after `return` statements...?)

**Comment by pnkfelix** [comment]
I have tagged this as NLL-diagnostics (rather than NLL-complete).

My reasoning: Even though this is an example where one might argue that NLL is causing the compiler to regress (and fail to compile previously compiled code, which would usually yield an NLL-complete issue), that failure to compile is a consequence of the use of `#![deny(warnings)]`, and our standard policy is that the compiler is free to add new warnings; cap-lints is meant to take care of downstream crates that have not been updated to account for any such warnings, as @nikomatsakis stated above.

**Comment by pnkfelix** [comment]
Triage: I don't think any hypothetical "fix" (if we make one) would be beta-backported. So the only question is what priority to assign.

With regards to prioritization... I'm torn between P-high or P-medium on this one. I think I'll wait to discuss it at the NLL team meeting tonight.

**Comment by pnkfelix** [comment]
Interestingly, the following example provided by @matthewjasper has the same behavior (of triggering the lint in question) on both the 2015 and 2018 editions ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=5c4511d874d839bdc65cc5f79cb75ffc))

```rust
#![deny(warnings)]  

#[allow(unreachable_code)]
pub fn sum_nan() {
    let mut v;
    v = 0;
    assert_eq!(v, 0);
    return;
    v = 1;
    assert_eq!(v, 1);
}

fn main() {}
```

**Comment by pnkfelix** [comment]
making executive decision to tag this as P-medium

**Comment by gnzlbg** [comment]
Thank you all!
